pytorch_lightning/utilities/fetching.py

```diff
@@ -0,0 +1,140 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from collections.abc import Iterable, Iterator
+from typing import Any, Generator, List, Optional, Tuple
+
+from torch.utils.data.dataloader import DataLoader
+
+from pytorch_lightning.trainer.supporters import CombinedLoader
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+class AbstractFetcher(ABC):
+
+    """
+    This class is used to control batch fetching flow.
+    """
+
+    @abstractmethod
+    def fetching_function(self) -> Generator:
+        pass
+
+    def __init__(
+        self,
+        prefetch_batches: int = 1,
+    ) -> None:
+        if not isinstance(prefetch_batches, int) or (isinstance(prefetch_batches, int) and prefetch_batches < 1):
+            raise MisconfigurationException("`prefetch_batches` should at least be 1.")
+
+        self.prefetch_batches = prefetch_batches
+        self.dataloader: Optional[Iterable]
+        self._has_setup: bool = False
+        self.reset()
+
+    def setup(self, dataloader: DataLoader, **kwargs) -> None:
+        if not isinstance(dataloader, (DataLoader, CombinedLoader)):
+            raise MisconfigurationException("The Fetcher should be setup with a ``dataloader``.")
+        self.dataloader = dataloader
+        self._has_setup = True
+
+    def add_batch(self, batch) -> None:
+        self.batches.append(batch)
+
+    def fetch_batch(self) -> Any:
+        return self.batches.pop(0)
+
+    @property
+    def loaders(self) -> List[DataLoader]:
+        if not self._has_setup:
+            raise MisconfigurationException("The Fetcher should be setup with a ``dataloader``.")
+        if isinstance(self.dataloader, CombinedLoader):
+            loaders = self.dataloader.loaders
+        else:
+            loaders = [self.dataloader]
+        return loaders
+
+    @property
+    def loader_iters(self) -> List[Iterator]:
+        if not self._has_setup:
+            raise MisconfigurationException("The Fetcher should be setup with a ``dataloader``.")
+        if isinstance(self.dataloader, CombinedLoader):
+            loader_iters = self.dataloader_iter.loader_iters
+        else:
+            loader_iters = [self.dataloader_iter]
+        return loader_iters
+
+    @property
+    def state(self) -> Any:
+        def collect_state(iterator: Iterator):
+            return iterator.state
+
+        return apply_to_collection(self.loader_iters, Iterator, collect_state)
+
+    def __iter__(self) -> Generator[Tuple[Any, bool], None, None]:
+        if self.dataloader is None:
+            raise MisconfigurationException("The iterate hasn't been provided. HINT: Did you call setup function ?.")
+        self.reset()
+        self.dataloader_iter = iter(self.dataloader)
```

Comment:
If storing this into a variable, rather already define that variable in `__init__`

Reply:
Done !